palace/fem/integ/curlcurlmass.cpp

```diff
@@ -0,0 +1,205 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "fem/integrator.hpp"
+
+#include <vector>
+#include <mfem.hpp>
+#include "fem/libceed/coefficient.hpp"
+#include "fem/libceed/integrator.hpp"
+
+#include "fem/qfunctions/curlcurlmass_qf.h"
+
+namespace palace
+{
+
+struct CurlCurlMassIntegratorInfo : public ceed::IntegratorInfo
+{
+  CurlCurlMassContext ctx;
+};
+
+namespace
+{
+
+CurlCurlMassIntegratorInfo
+InitializeIntegratorInfo(const mfem::ParFiniteElementSpace &fespace,
+                         const mfem::IntegrationRule &ir, const std::vector<int> &indices,
+                         bool use_bdr, mfem::Coefficient *Qc, mfem::VectorCoefficient *VQc,
+                         mfem::MatrixCoefficient *MQc, mfem::Coefficient *Qm,
+                         mfem::VectorCoefficient *VQm, mfem::MatrixCoefficient *MQm,
+                         std::vector<ceed::QuadratureCoefficient> &coeff)
+{
+  MFEM_VERIFY(fespace.GetVDim() == 1,
+              "libCEED interface for CurlCurlMassIntegrator does not support vdim > 1!");
+
+  CurlCurlMassIntegratorInfo info = {{0}};
+
+  mfem::ParMesh &mesh = *fespace.GetParMesh();
+  info.ctx.dim = mesh.Dimension() - use_bdr;
+  info.ctx.space_dim = mesh.SpaceDimension();
+  info.ctx.curl_dim = (info.ctx.dim < 3) ? 1 : info.ctx.dim;
+
+  info.trial_op = ceed::EvalMode::InterpAndCurl;
+  info.test_op = ceed::EvalMode::InterpAndCurl;
+  info.qdata_size = (info.ctx.curl_dim * (info.ctx.curl_dim + 1)) / 2 +
+                    (info.ctx.dim * (info.ctx.dim + 1)) / 2;
+
+  MFEM_VERIFY((Qc || VQc || MQc) && (Qm || VQm || MQm),
+              "libCEED CurlCurlMassIntegrator requires both a "
+              "curl-curl and a mass integrator coefficient!");
+  if (Qc)
+  {
+    coeff.emplace_back();
+    ceed::InitCoefficient(*Qc, mesh, ir, indices, use_bdr, coeff.back());
```

Comment:
The performance benefit is very small since setup is done once for all, and most of the time you care about allowing non-constant coefficients on both. Hence this integrator allowing only this more general case.